### ISSUE TITLE
Documentation for client authors incorporating STS

### ISSUE THREAD

**Issue by thwg** [ircv3/ircv3-specifications#211]
STS could benefit from additional documentation outside of the specification to help client authors meet the requirement:

> Clients MUST use sufficient certificate verification to ensure the connection is secure without any errors.

This is a continuation of an off-topic discussion in #209.

------

Feel free to correct if I'm mistaken on any parts, as I'm not an expert w.r.t. X.509.

This specification provides server rules to provide info that the client might want (the secure port and instructions to use TLS in the future), but the problem was never with the info the servers provided -- the problem is reliably verifying the identity of the server against a trusted root CA. The specification mentions it, but only briefly:

> Clients MUST use sufficient certificate verification to ensure the connection is secure without any errors.

Given my rusty understanding, but based on what web browsers do, in the future we would do something to the effect of:

- Collect a list of the Common trusted root CAs (e.g. Debian's ca-certificates)
- Keep an up-to-date list of CRLs (maybe OCSP nowadays?)
- Distribute them to mirrors (PGP signed)
- Include an update mechanism in client, and update regularly from mirrors
- Ship the public PGP key with the client distribution to verify the authenticity of each update
- Keep an additional list of certificates that the user has personally trusted.

If this were easy then we would all be TLS-preferred in the default servers list a long time ago. I think some good advice for what "use sufficient certificate verification" means would be helpful to most client authors. Not included with the spec, but perhaps placed the documentation/ directory.

**Comment by thwg** [comment]
> I'd like to assume
> that people who install clients from source tarball are knowledgeable enough to research TLS error messages or at least read the "Dependencies" list in the client's README.TXT

This is what was being done previously for all users.  Users configure gnutls or openssl and spawn an inferior process.  It won't protect them from bad default behavior though.  I'm not aware of any portable library that does this, but would be interested if there is one.  This is why TLS has not been the default.

> Doesn't your TLS library already have a compile-time default provided by the distro packager?

To my knowledge, it is optional and non-default:

>  Please note that Debian can neither confirm nor deny whether the
> certificate authorities whose certificates are included in this package
> have in any way been audited for trustworthiness or RFC 3647 compliance.
> Full responsibility to assess them belongs to the local system
> administrator.

The link you noted is a good example ^^.

> How do Hexchat and KVIrc handle it?

Not sure.  Do they verify identity by default?

**Comment by jwheare** [comment]
Here's what Chrome does: https://www.chromium.org/Home/chromium-security/root-ca-policy, using the OS root store on OS X and Windows, and on Linux using a Mozilla root store via a library.

**Comment by grawity** [comment]
> This is what was being done previously for all users. Users configure gnutls or openssl and spawn an inferior process. It won't protect them from bad default behavior or lack of installed common CAs though. I'm not aware of any portable library that does this, but would be interested if there is one. This is why TLS ports have not been in the default server list -- unless we are shipping our own cert db and CRLs.

Huh? What does this have to do with spawning a separate process? I'm talking about the TLS libraries themselves, not about gnutls-cli stuff.

> To my knowledge, it is optional and non-default:

>> Please note that Debian can neither confirm nor deny whether the
>> certificate authorities whose certificates are included in this package
>> have in any way been audited for trustworthiness or RFC 3647 compliance.
>> Full responsibility to assess them belongs to the local system
>> administrator.

Um, that quote talks about the package contents' _validity_ (i.e. Debian just blindly pulls stuff from Mozilla), it says nothing about when the package itself gets installed.

**Comment by thwg** [comment]
> Huh? What does this have to do with spawning a separate process? I'm talking about the TLS libraries themselves, not about gnutls-cli stuff.

The client is old and written in Emacs Lisp; I'm updating it to support IRCv3 standards.  TLS is implemented by piping to a gnutls-cli or openssl inferior process.  This was due to limitations in dynamic linking at the time, but it can be updated to use libraries now.

> Um, that quote talks about the package contents' validity (i.e. Debian just blindly pulls stuff from Mozilla), it says nothing about when the package itself gets installed.

Yes, what I mean to say is that gnutls and associated libraries may be installed on some platforms without pulling in a sane set of common CAs (it's up to the administrator, Debian does not enforce this).  But more importantly, there would previously always be a manual steps if they wanted to use TLS, so they can read their distro documentation, install related packages, and set the cafile (wherever it may be) / appropriate options.

Most users just use the default, so I don't want to silently convert users to 'secure' connections that aren't secure, or break their existing setups with non-working ones.  The alternative is to do the same thing that Firefox and Chromium on Linux do, which is tools to bundle / manage some of their own certs and CRLs.  I was thinking that some documentation on how to do this would be helpful for myself and other client authors, but maybe it's just a localized problem? :-)

I did get the opportunity test the latest HexChat with a self-signed cert which produces an error, which does validate the identity correctly, but connects anyways.  This may be a user error/configuration though:

    * * Verify E: self signed certificate in certificate chain.? (19) -- Ignored
    * Connected. Now logging in.

> Here's what Chrome does: https://www.chromium.org/Home/chromium-security/root-ca-policy, using the OS root store on OS X and Windows, and on Linux using a Mozilla root store via a library.

Thank you for the link.  I'll look further into NSS -- it looks promising.

**Comment by thwg** [comment]
Thinking about this issue some more, I'll admit that it's probably a localized problem to my IRC client, and not an issue for most people.  Either implementing a dialog to have the user so that the user must configure their client before connecting to a STS enabled server or using some tools from NSS would both work for my scenario.

I'll close this issue.

**Comment by TingPing** [comment]
> I did get the opportunity test the latest HexChat with a self-signed cert, which does validate the identity correctly, but connects anyways. This may be a user/configuration error though:

If you use the network list HexChat never ignores certs without the user explicitly telling it to do so.

**Comment by jwheare** [comment]
Can you delete your comments from #209 to clean it up please.

**Comment by thwg** [comment]
@TingPing: Ah, makes sense. I was using a test network, not on the network list.

@jwheare: Yes, sorry, I've removed my posts from that issue.

Just for posterity, I've realized I explained my issue badly, but the short of it was:

Users install outside of their package manager and may have zero or more of the programs installed:
  - OpenSSL
  - GnuTLS
  - Common CA certificates

Previously only users who understood TLS would use it, so they could reasonably understand how to set it up.  When enabling TLS for all users I would like to replicate the behavior of major web browsers (e.g. they don't ask the user to go download other these libraries to connect to HTTPS).  It would need to work on all major operating systems, and handle things like certs and revocation lists.

Mozilla's NSS seems to be a good tool for the tackling the problem.  I'll look further into packaging it with the program.

Apologies for the noise on your tracker, this was not the correct place for my issue.
